### PR TITLE
feat: add Chapter 5

### PR DESCRIPTION
**Pull request by skylee03**

### CODE REVIEW

File: docs/05-trees.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    Empty => false // empty tree implies we are getting deepest
+    Node(value, left, right) => // otherwise, search in subtrees
+      value == target || dfs_search(target, left) || dfs_search(target, right)
+  }
+}
+```
+
+As we introduced earlier, this is a traversal based on structural recursion. We first handle the base case, i.e., when the tree is empty, as shown in the third line. In this case, we haven't found the value we're looking for, so we return `false`. Then, we handle the recursive case. For a node, we check if its value is the desired result, as shown in line 5. If we find it, the result is `true`. Otherwise, we continue to traverse the left and right subtrees alternately, if either we find it in left subtree or right subtree will the result be `true`. In the current binary tree, we need to traverse both the left and right subtrees to find the given value. The binary search tree introduced later will optimize this process. The only differences between preorder, inorder, and postorder searches is the order of operations on the current node, the left subtree search, and the right subtree search.
+
+### Short-Circuit Evaluation of Booleans
```
Comment (peter-jerry-ye) [inline docs/05-trees.md]: Move this part to previous chapters

File: docs/05-trees.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+The key to maintaining balance in a binary balanced tree is that when the tree becomes unbalanced, we can rearrange the tree to regain balance. The insertion and deletion of AVL trees are similar to standard binary search trees, except that AVL trees perform adjustments after each insertion or deletion to ensure the tree remains balanced. We add a height attribute to the node definition. 
+
+```moonbit no-check
+enum AVLTree {
+  Empty
+  // current value, left subtree, right subtree, height
+  Node(Int, AVLTree, AVLTree, Int) 
+}
+fn create(value: Int, left: AVLTree, right: AVLTree) -> AVLTree
```
Comment (peter-jerry-ye) [inline docs/05-trees.md]: Maybe it's an opportunity to introduce labelled argument and default value.
```
fn create(value : Int, ~left : AVLTree = Empty, ~right : AVLTree = Empty) -> AVLTree
```

File: docs/05-trees.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+  } else { ... }
+}
+```
+
+Here is a snippet of code for a balanced tree. You can easily complete the code once you understand what we just discussed. We first determine if a tree is unbalanced, by checking if the height difference between the two subtrees exceeds a specific value and which side is higher. After determining this, we perform a rebalancing operation. At this point, the root node we pass in is $z$, and the higher side after pattern matching is $y$. Then, based on the comparison of the heights of $y$'s two subtrees, we further determine whether $x$ is $y$ and $z$'s same side or between $y$ and $z$, as shown in line 6. Afterwards, we recombine based on the scenarios discussed earlier, as shown in lines 6 and 9. Taking insertion of an element as an example: 
+
+```moonbit no-check
+fn add(tree: AVLTree, value: Int) -> AVLTree {
+  match tree {
+    Node(v, left, right, _) as t => {
```
Comment (peter-jerry-ye) [inline docs/05-trees.md]: Maybe we can introduce about label in enum in #22 and rewrite this part as something like
```
match tree {
  Node(v, ..) as t => {
    if value < v { balance(add(t.left, value), v, t.right) } else { ... }
  }
}
```
where
```
enum AVLTree {
  Empty
  Node(Int, ~left: AVLTree, ~right: AVLTree, ~height: Int)
}
```

File: docs/05-trees.md (hunk 2 above)
Comment (skylee03) [reply docs/05-trees.md]: I will introduce them in Chapter 3.

File: docs/05-trees.md (hunk 1 above)
Comment (skylee03) [reply docs/05-trees.md]: I will move this part to Chapter 2.

**Review by peter-jerry-ye — APPROVED** [review]

**Pull request merged**